drivers/usb/gadget/xaptum/f_psock_gadget.c

```diff
@@ -89,21 +100,91 @@ void f_psock_gadget_work_handler( struct work_struct *work )
  * USB DESCRIPTOR DEFINITIONS
  ***************************************************************************/
 
+/* Control Endpoint Descriptions. 
+ * These descriptors do not vary by speed setting becuase FS, HS and SS all 
+ * have lengths >=64 bytes, and SCM protocol standards limit control 
+ * transfer packets to 64 bytes.
+ */
+static struct usb_endpoint_descriptor
+f_psock_fs_ctrl_sink_desc = {
+	.bLength         = USB_DT_ENDPOINT_SIZE,
+	.bDescriptorType = USB_DT_ENDPOINT,
+
+	.bEndpointAddress = USB_DIR_OUT,
+	.bmAttributes    = USB_ENDPOINT_XFER_INT,
+	.wMaxPacketSize  = cpu_to_le16(MAX_CTRL_PACKET_SIZE),
+	.bInterval	 = 32,
+};
+
+static struct usb_endpoint_descriptor
+f_psock_fs_ctrl_source_desc  = {
+	.bLength         = USB_DT_ENDPOINT_SIZE,
+	.bDescriptorType = USB_DT_ENDPOINT,
+
+	.bEndpointAddress = USB_DIR_IN,
+	.bmAttributes    = USB_ENDPOINT_XFER_INT,
+	.wMaxPacketSize  = cpu_to_le16(MAX_CTRL_PACKET_SIZE),
+	.bInterval 	 = 32,
+};
+
+static struct usb_endpoint_descriptor
+f_psock_hs_ctrl_sink_desc = {
+	.bLength         = USB_DT_ENDPOINT_SIZE,
+	.bDescriptorType = USB_DT_ENDPOINT,
+
+	.bEndpointAddress = USB_DIR_OUT,
+	.bmAttributes    = USB_ENDPOINT_XFER_INT,
+	.wMaxPacketSize  = cpu_to_le16(MAX_CTRL_PACKET_SIZE),
+	.bInterval	 = USB_MS_TO_HS_INTERVAL(32),
+};
+static struct usb_endpoint_descriptor
+f_psock_hs_ctrl_source_desc = {
+	.bLength         = USB_DT_ENDPOINT_SIZE,
+	.bDescriptorType = USB_DT_ENDPOINT,
+
+	.bEndpointAddress = USB_DIR_IN,
+	.bmAttributes    = USB_ENDPOINT_XFER_INT,
+	.wMaxPacketSize  = cpu_to_le16(MAX_CTRL_PACKET_SIZE),
+	.bInterval 	 = USB_MS_TO_HS_INTERVAL(32),
+};
+
+
+static struct usb_endpoint_descriptor
+f_psock_ss_ctrl_sink_desc = {
+	.bLength         = USB_DT_ENDPOINT_SIZE,
+	.bDescriptorType = USB_DT_ENDPOINT,
+
+	.bEndpointAddress = USB_DIR_OUT,
+	.bmAttributes    = USB_ENDPOINT_XFER_INT,
+	.wMaxPacketSize  = cpu_to_le16(MAX_CTRL_PACKET_SIZE),
+	.bInterval	 = USB_MS_TO_HS_INTERVAL(32),
+};
+static struct usb_endpoint_descriptor
+f_psock_ss_ctrl_source_desc = {
+	.bLength         = USB_DT_ENDPOINT_SIZE,
+	.bDescriptorType = USB_DT_ENDPOINT,
+
+	.bEndpointAddress = USB_DIR_IN,
+	.bmAttributes    = USB_ENDPOINT_XFER_INT,
+	.wMaxPacketSize  = cpu_to_le16(MAX_CTRL_PACKET_SIZE),
+	.bInterval 	 = USB_MS_TO_HS_INTERVAL(32),
+};
+
 /*
  * usb interface descriptor
  */
 static struct usb_interface_descriptor psock_intf = {
 	.bLength = sizeof(psock_intf),
 	.bDescriptorType = USB_DT_INTERFACE,
-	.bNumEndpoints = 2,
+	.bNumEndpoints = 4,
```

Comment:
This was the cause of a crash. Using the device on hardware circumvented the crash and the host (Advantech gateway) was able to report a mismatch in the descriptor count that lead me to this...

Reply:
nice!